service/src/ConstConfigIOGroup.hpp

```diff
@@ -0,0 +1,174 @@
+            void restore_control(void) override;
+            /// @return  a function that should be used when aggregating the given signal.
+            ///
+            /// @see geopm::Agg
+            std::function<double(const std::vector<double> &)> agg_function(const std::string &signal_name) const override;
+            /// @return  a function that should be used when formatting the given signal.
+            ///
+            /// @see geopm::Agg
+            std::function<std::string(double)> format_function(const std::string &signal_name) const override;
+            /// @return a string description for signal_name, if defined.
+            std::string signal_description(const std::string &signal_name) const override;
+            /// @throws geopm::Exception there are no controls supported by the ConstConfigIOGroup
+            std::string control_description(const std::string &control_name) const override;
+            /// @return M_SIGNAL_BEHAVIOR_CONSTANT 
+            int signal_behavior(const std::string &signal_name) const override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            ///
+            /// @param save_path this argument is ignored
+            void save_control(const std::string &save_path) override;
+            /// @brief Does nothing; this IOGroup does not provide any controls.
+            ///
+            /// @param save_path this argument is ignored
+            void restore_control(const std::string &save_path) override;
+            /// @return the name of the IOGroup
+            std::string name(void) const override;
+            /// @return the name of the plugin to use when this plugin is
+            ///         registered with the IOGroup factory
+            ///
+            /// @see geopm::PluginFactory
+            static std::string plugin_name(void);
+            /// @return a pointer to a new ConstConfigIOGroup object
+            ///
+            /// @see geopm::PluginFactory
+            static std::unique_ptr<IOGroup> make_plugin(void);
+
+        private:
+            struct m_signal_desc_s {
+                json11::Json::Type type;
+                bool required;
+            };
+
+            struct m_signal_info_s {
+                m_signal_info_s(
+                    int units,
+                    int domain,
+                    std::function<double(const std::vector<double> &)> agg,
+                    std::string desc,
+                    std::vector<double> values):
+
+                    units(units),
+                    domain(domain),
+                    agg_function(agg),
+                    description(desc),
+                    values(values)
+                {}
+
+                int units;
+                int domain;
+                std::function<double(const std::vector<double> &)> agg_function;
+                std::string description;
+                std::vector<double> values;
+            };
```

Comment:
I think this can be simplified a bit.  I think there are some redundant/unused fields with the current implementation.  Will this work?

```suggestion            
            struct m_signal_info_s
            {                
                int domain;
                int units;
                std::function<double(const std::vector<double> &)> agg_function;
                std::string description;
                std::vector<double> values;
            };
```

Reply:
Wait, it seems your suggestion is to get rid of the constructor (i.e. I'm not seeing changes related to potentially redundant fields). The constructor is needed as a convenience when calling ``std::make_shared<m_signal_info_s>(...)``.

Reply:
Is there a reason that ```m_signal_available``` is defined that way (having a shared pointer to the signal info struct?  In all other IOGroups I looked at, the definition is just:
```
std::map<std::string, m_signal_info_s> m_signal_available;
```

I don't see the need to make ```m_signal_info_s``` a shared_ptr.  It is never passed outside this class.

Take a look at how ```m_signal_available``` is defined and used in the ```MSRIOGroup```.